### PR TITLE
change search ux

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Before

Display the searched note in a popup window.
## After

Display the searched notes in the normal note list.

### CODE REVIEW

**Comment by huettenhain** [comment]
@sosukesuzuki this looks awesome.

**Comment by kazup01** [comment]
Awesome!

File: browser/main/NoteList/index.js [hunk 1]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
     })
   }
 
+  getSearchNotes () {
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: To begin with my review, I'd like you to move this method to a module such as `lib/search` or something. But it's ok, I'll start my review :)

File: browser/main/NoteList/index.js [hunk 2]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
+        let regExp = new RegExp(_.escapeRegExp(tag), 'i')
+        notes = notes
+          .filter((note) => {
+            if (!_.isArray(note.tags)) return false
+            return note.tags.some((_tag) => {
+              return _tag.match(regExp)
+            })
+          })
+      } else if (block.match(/^!.+/)) {
+        let block = block.match(/^!(.+)/)[1]
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: This line contains a bug even in production :fearful:

File: browser/main/NoteList/index.js [hunk 3]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
+          }
+          if (note.type === 'SNIPPET_NOTE') {
+            return note.description.match(regExp)
+          } else if (note.type === 'MARKDOWN_NOTE') {
+            return note.content.match(regExp)
+          }
+          return false
+        })
+      }
+  })
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: 😎

File: browser/main/NoteList/index.js [hunk 4]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
   }
 
+  getSearchNotes () {
+    let { data } = this.props
+    let search = document.getElementsByClassName('TopBar__control-search-input___browser-main-TopBar-')[0].childNodes[0].value
+    let notes = data.noteMap.map((note) => note)
+    if (search.trim().length === 0) return []
+    let searchBlocks = search.split(' ')
+    searchBlocks.forEach((block) => {
+      if (block.match(/^!#.+/)) {
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: You can delete these condition starting from `/^!.+/`.

File: browser/main/NoteList/index.js (hunk 2 above)
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: You can delete this condition too.

File: browser/main/NoteList/index.js [hunk 5]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
+            return true
+          }
+          if (note.type === 'SNIPPET_NOTE') {
+            return !note.description.match(regExp)
+          } else if (note.type === 'MARKDOWN_NOTE') {
+            return !note.content.match(regExp)
+          }
+          return false
+        })
+      } else if (block.match(/^#.+/)) {
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: You should make a new method named such as `findByTag()`.

File: browser/main/NoteList/index.js [hunk 6]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
+        let tag = block.match(/#(.+)/)[1]
+        let regExp = new RegExp(_.escapeRegExp(tag), 'i')
+        notes = notes
+          .filter((note) => {
+            if (!_.isArray(note.tags)) return false
+            return note.tags.some((_tag) => {
+              return _tag.match(regExp)
+            })
+          })
+      } else {
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: Also, you can make a new method such ad `findByWord()`.

File: browser/main/NoteList/index.js [hunk 7]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
...
+            return _tag.match(regExp)
+          })) {
+            return true
+          }
+          if (note.type === 'SNIPPET_NOTE') {
+            return note.description.match(regExp)
+          } else if (note.type === 'MARKDOWN_NOTE') {
+            return note.content.match(regExp)
+          }
+          return false
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: I assume this below is ideal form.

```
if (isTag(block)) {
  let tag = block.match(/#(.+)/)[1]
  let regExp = new RegExp(_.escapeRegExp(tag), 'i')
  notes = findByTag(notes, regExp)
} else {
  let regExp = new RegExp(_.escapeRegExp(block), 'i')
  notes = findByWord(notes, regExp)
}
```

Comment (asmsuechan) [inline browser/main/NoteList/index.js]: And also, we have to contemplate to decide these specifications.

Comment (asmsuechan) [inline browser/main/NoteList/index.js]: the magic number is not good 🙅

File: browser/main/NoteList/index.js [hunk 8]
```diff
@@ -349,6 +353,71 @@ class NoteList extends React.Component {
     })
   }
 
+  getSearchNotes () {
+    let { data } = this.props
+    let search = document.getElementsByClassName('TopBar__control-search-input___browser-main-TopBar-')[0].childNodes[0].value
```
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: The class name depends on webpack, so you have to change it. If I were you, I name the element such as `searchInput`.

File: browser/main/NoteList/index.js (hunk 1 above)
Comment (asmsuechan) [inline browser/main/NoteList/index.js]: I assume `searchFromNotes()` is more appropriate.

Comment (sosukesuzuki) [inline browser/main/NoteList/index.js]: Thank you for commenting! I think so, too. I'll change it.

File: browser/main/NoteList/index.js (hunk 8 above)
Comment (sosukesuzuki) [inline browser/main/NoteList/index.js]: Thank you for telling it me! I'll add className 'searchInput' to `input` in `Topbar/index.js`.

File: browser/main/NoteList/index.js (hunk 7 above)
Comment (sosukesuzuki) [inline browser/main/NoteList/index.js]: Thank you for commenting! That is so simple and so good. I'll change to it.

File: browser/lib/search.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  searchBlocks.forEach((block) => {
+    if (block.match(/^#.+/)) {
+      const tag = block.match(/#(.+)/)[1]
+      notes = findByTag(notes, tag)
+    } else {
+      notes = findByWord(notes, block)
+    }
+  })
+  return notes
+}
```
Comment (asmsuechan) [inline browser/lib/search.js]: Thank you for changing them into the module :smile:
And probably, I made a mistake :fearful:. I assume arranged interface is better (regarding `findByTag()` and `findByWord`). So, you should change `findByTag()`. It's better that to cut out the tag from a block inside the method.

```
searchBlocks.forEach((block) => {
  if (block.match(/^#.+/)) {
     notes = findByTag(notes, block)
  } else {
     notes = findByWord(notes, block)
  }
})
```

And also, you can make a method such as `findByTagOrWord(block)` to combine the methods for finding. (I'm not sure the argument name (block) is better or not :face_with_thermometer:)

Comment (sosukesuzuki) [inline browser/lib/search.js]: Thank you for commenting! Certainly it is simple and easy to understand.

**Comment by kazup01** [comment]
🎉🎉🎉🎉🎉🎉🎉

**Pull request merged**